internal/server/grants_test.go

```diff
@@ -0,0 +1,314 @@
+		assert.NilError(t, err)
+
+		req, err := http.NewRequest(http.MethodPost, "/v1/identities", &buf)
+		assert.NilError(t, err)
+		req.Header.Add("Authorization", "Bearer "+adminAccessKey(srv))
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+		assert.Equal(t, resp.Code, http.StatusCreated, resp.Body.String())
+		respObj := &api.CreateIdentityResponse{}
+		err = json.Unmarshal(resp.Body.Bytes(), respObj)
+		assert.NilError(t, err)
+		return respObj.ID
+	}
+
+	createGrant := func(t *testing.T, subject uid.PolymorphicID, privilege string) {
+		t.Helper()
+		var buf bytes.Buffer
+		body := api.CreateGrantRequest{
+			Subject:   subject,
+			Privilege: privilege,
+			Resource:  "res1",
+		}
+		err := json.NewEncoder(&buf).Encode(body)
+		assert.NilError(t, err)
+
+		req, err := http.NewRequest(http.MethodPost, "/v1/grants", &buf)
+		assert.NilError(t, err)
+		req.Header.Add("Authorization", "Bearer "+adminAccessKey(srv))
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+		assert.Equal(t, resp.Code, http.StatusCreated, resp.Body.String())
+	}
+
+	createGroup := func(t *testing.T, name string, users ...uid.ID) uid.ID {
+		t.Helper()
+		group := &models.Group{Name: name}
+		for _, user := range users {
+			iden := models.Identity{Model: models.Model{ID: user}}
+			group.Identities = append(group.Identities, iden)
+		}
+
+		err := data.CreateGroup(srv.db, group)
+		assert.NilError(t, err)
+		return group.ID
+	}
+
+	idInGroup := createID(t, "inagroup@example.com")
+	idOther := createID(t, "other@example.com")
+
+	createGrant(t, uid.NewIdentityPolymorphicID(idInGroup), "custom1")
+	createGrant(t, uid.NewIdentityPolymorphicID(idOther), "custom2")
+
+	groupID := createGroup(t, "humans", idInGroup)
+
+	token := &models.AccessKey{
+		IssuedFor:  idInGroup,
+		ProviderID: data.InfraProvider(srv.db).ID,
+		ExpiresAt:  time.Now().Add(10 * time.Second),
+	}
+
+	accessKey, err := data.CreateAccessKey(srv.db, token)
+	assert.NilError(t, err)
+
+	admin, err := data.GetIdentity(srv.db, data.ByName("admin"))
+	assert.NilError(t, err)
+
+	type testCase struct {
+		urlPath  string
+		setup    func(t *testing.T, req *http.Request)
+		expected func(t *testing.T, resp *httptest.ResponseRecorder)
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		req, err := http.NewRequest(http.MethodGet, tc.urlPath, nil)
+		assert.NilError(t, err)
+		req.Header.Set("Authorization", "Bearer "+accessKey)
+
+		if tc.setup != nil {
+			tc.setup(t, req)
+		}
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+
+		tc.expected(t, resp)
+	}
+
+	testCases := map[string]testCase{
+		"not authenticated": {
+			urlPath: "/v1/grants?subject=i:" + idOther.String(),
+			setup: func(t *testing.T, req *http.Request) {
+				req.Header.Del("Authorization")
+			},
+			expected: func(t *testing.T, resp *httptest.ResponseRecorder) {
+				assert.Equal(t, resp.Code, http.StatusUnauthorized)
+			},
+		},
+		"not authorized, wrong identity": {
```

Comment:
```suggestion
		"authorized, wrong identity": {
```

Reply:
actually... I think these are confusing authorized/authenticated. It's not clear to me what they should say